[New Content] - Overture Guides (#385)
overture-stack/website: markdown/documentation/guides/deployment/datamanagement.md

---
title: Deployment Guide
---

# Data Management and Storage Setup

> Song, Score and their dependent services (Postgres, Minio, Kafka, and Zookeeper) will be set up next, together these services will manage, track and store all our data on the backend. 

Reviewer: Can remove zookeeper from list if that is no longer used.

> Song, Score and their dependent services (Postgres, Minio, Kafka, and Zookeeper) will be set up next, together these services will manage, track and store all our data on the backend. 

Reviewer: ```suggestion
Song, Score and their dependent services (Postgres, Minio, and Kafka) will be set up next. These services will manage, track and store all our data on the backend. 
```

## Setting up Object Storage

> The file transfer service Score is compatible with any S3 storage provider; for simplicity, we will use the open-source Object Storage provider Minio for this setup. 

Reviewer: ```suggestion
The file transfer service Score is compatible with any S3 storage provider; for simplicity, we will use the open-source object storage provider [Minio](https://min.io/) for this setup. 
```

1. **Run Minio:** Use the following command to pull and run the Minio docker container

```bash
docker run -d --name minio \
  -p 9000:9000 \
  -e MINIO_ACCESS_KEY=admin \
  -e MINIO_SECRET_KEY=admin123 \
  -v $./persistentStorage/data-minio:/data \
  minio/minio:RELEASE.2018-05-11T00-29-24Z \
  server /data
```

> 2. **Run the Minio client:** This command will start up a minio-client set up to create an object and state bucket

Reviewer: I think this action is not about "running the client" it is about "create buckets to store data". We can mention that we are using the client, but thats secondary to the point:

```suggestion
2. **Create buckets in Minio:** Run the following command to create two buckets in the running Minio server using the Minio CLI - this will create an **object** bucket to store for uploaded files and a **state** bucket for metadata files managed by Score.
```

```bash
docker run --name minio-client \
  --entrypoint /bin/sh \
  minio/mc -c \
  '/usr/bin/mc alias set myminio http://host.docker.internal:9000 admin admin123 && \
  /usr/bin/mc mb myminio/state && \
  /usr/bin/mc mb myminio/object && \
  /usr/bin/mc mb myminio/object/data && \
  echo "" > /tmp/heliograph && \
  /usr/bin/mc put /tmp/heliograph myminio/object/data/heliograph && \
  rm /tmp/heliograph && \
  exit 0;'
```

> You should now be able to access the minio console from the `localhost:9000`

Reviewer: from the browser at `localhost:9000` 

or 

from `localhost:9000`

<details>

  <summary><b>For more details, click here</b></summary>

<br></br>

### Minio Image

> - The `-v $./persistentStorage/data-minio:/data` will store our uploaded mock file data in our local directory

Reviewer: ```suggestion
- The `-v $./persistentStorage/data-minio:/data` configures Minio to store data in our local filesystem instead of in the docker container. Files you upload to Minio will be stored at the path `./persistentStorage/data-minio`.
```

- **Object Bucket:** `mb myminio/object` creates another bucket named "object". The "object" bucket is intended for storing the actual content objects, such as VCFs, BAMs, etc. 

- **Data directory & Heliograph File:** The `put` command seeds an empty' heliograph' file within the object storage data folder. Score uses this dummy file to test that the server can successfully communicate with the storage provider and that your client can successfully retrieve files from it, too.

---

</details>

<br></br>

## Setting up the Song Database

1. **Run PostgreSQL:** Use the following command to pull and run the PostgreSQL docker container

```bash
docker run --name song-db --network db-network  -e POSTGRES_USER=admin -e POSTGRES_PASSWORD=admin123 -e POSTGRES_DB=songDb -v./persistentStorage/data-song-db:/var/lib/postgresql/data -d postgres:11.1
```
<details>

  <summary><b>For more details, click here</b></summary>

<br></br>

- This command runs a postgres image named `song-db` on the `db-network` with the username `admin`, password `admin123` and a database within it called `songDb`.

> - We are including a defined persistent volume `-v ./persistentStorage/song-db-data:/var/lib/postgresql/data`. This volume will be a folder generated at runtime to serve as persistent storage, meaning the data in your database will persist regardless of the container's status, in this case, located in the root of the directory where you run the container

Reviewer: - Needs to end in a period.
- Probably want to use the term "local filesystem" and mention how this stores the postgres at the specific path `./persistentStorage/song-db-data`.

See similar comment about Minio persistent storage.

---
</details>

<br></br>

<Note title="Bringing it together">By setting up Minio alongside PostgreSQL, we have created an environment capable of handling both relational and object data storage, simulating a backend infrastructure similar to what you'd find in a cloud-based application.</Note>

## Running Zookeeper and Kafka

1. **Run Zookeeper:** Use the following command to pull and run the Zookeeper docker container

```bash
docker run -d --name zookeeper \
  -p 2181:2181 \
  -e ZOOKEEPER_CLIENT_PORT=2181 \
  -e ZOOKEEPER_TICK_TIME=2000 \
  confluentinc/cp-zookeeper:7.6.1

> ```

Reviewer: If you are removing zookeeper then this section can be removed and the header can be updated.

> 2. **Run the Kafka:** Use the following command to pull and run the Kafka docker container

Reviewer: "Run kafka" or "the Kafka server"

```bash
docker run -d --name kafka \
  --platform linux/amd64 \
  -p 29092:29092 \
  -p 9092:9092 \
  -e KAFKA_BROKER_ID=1 \
  -e KAFKA_ZOOKEEPER_CONNECT=zookeeper:2181 \
  -e KAFKA_LISTENER_SECURITY_PROTOCOL_MAP=PLAINTEXT:PLAINTEXT,PLAINTEXT_HOST:PLAINTEXT \
  -e KAFKA_ADVERTISED_LISTENERS=PLAINTEXT://kafka:29092,PLAINTEXT_HOST://kafka:9092 \
  -e KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR=1 \
  -e CONFLUENT_METRICS_REPORTER_BOOTSTRAP_SERVERS=kafka:9092 \
  -e CONFLUENT_METRICS_REPORTER_ZOOKEEPER_CONNECT=zookeeper:2181 \
  confluentinc/cp-kafka:5.4.0

```

> <Note title="What is this for?">Kafka handles the distribution and processing of data streams, while Zookeeper manages the coordination and configuration aspects of the Kafka cluster. Together they serve as the backbone messaging system for Song, enabling asynchronous communication between Song and Maestro ensuring efficient and reliable job execution, queuing, and processing..</Note>

Reviewer: two periods at end..

## Running Song

> Song is our data cataloging system. It will automate submission validations and track and manage all our metadata and file data.

Reviewer: Song doesn't really "automate"... it will:
- store and validate the metadata for uploaded files
- allow management of the published status of files
- notify downstream services about new files and updates to existing files

1. **Create a env file:** Create a file named `.env.song` with the following content:

```bash
# ==============================
# Song Environment Variables
# ==============================

# Spring Run Profiles
SPRING_PROFILES_ACTIVE=prod,secure,kafka
# Flyway variables
SPRING_FLYWAY_ENABLED=true
# Song Variables
ID_USELOCAL=true
SCHEMAS_ENFORCELATEST=true
# Score Variables
SCORE_URL=http://score:8087

> SCORE_ACCESSTOKEN=68fb42b4-f1ed-4e8c-beab-3724b99fe528

Reviewer: Maybe we shouldn't put a placeholder here, just leave it empty for now. This value is meaningless.

# Keycloak Variables
AUTH_SERVER_PROVIDER=keycloak
AUTH_SERVER_CLIENTID=dms
AUTH_SERVER_CLIENTSECRET=t016kqXfI648ORoIP5gepqCzqtsRjlcc
AUTH_SERVER_TOKENNAME=apiKey
AUTH_SERVER_KEYCLOAK_HOST=http://keycloak:8080
AUTH_SERVER_KEYCLOAK_REALM=myrealm
AUTH_SERVER_SCOPE_STUDY_PREFIX=STUDY.
AUTH_SERVER_SCOPE_STUDY_SUFFIX=.WRITE
AUTH_SERVER_SCOPE_SYSTEM=song.WRITE
SPRING_SECURITY_OAUTH2_RESOURCESERVER_JWT_JWK_SET_URI=http://keycloak:8080/realms/myrealm/protocol/openid-connect/certs
AUTH_SERVER_INTROSPECTIONURI=http://keycloak:8080/realms/myrealm/apikey/check_api_key/
# Postgres Variables
SPRING_DATASOURCE_URL=jdbc:postgresql://song-db:5432/songDb?stringtype=unspecified
SPRING_DATASOURCE_USERNAME=admin
SPRING_DATASOURCE_PASSWORD=admin123
# Kafka Variables
SPRING_KAFKA_BOOTSTRAPSERVERS=http://kafka:9092
SPRING_KAFKA_TEMPLATE_DEFAULTTOPIC=song-analysis
# Swagger Variable
SWAGGER_ALTERNATEURL=/swagger-api
```

> <Note title="Score Access Token">We will update our `SCORE_ACCESSTOKEN` value after portal deployment. Once Stage, our portal UI is deployed, we can more easily generate an admin API key with the appropriate credentials</Note>

Reviewer: Won't the access token be a unique key for song that is generated in Keycloak, not Stage?

When using the Ego backend (that I am more familiar with) we used Application credentials (application ID and Secret) and the access token was generated on the fly after making a request to Ego. Is that not the case here?

Author: I can be done, either way, after some troubleshooting and on @leoraba recommendation went with Access Key,  tagged him as so he can elaborate, I'll add it to the final note on the Quickstart PR before merging to main

Reviewer 2: To clarify SCORE_ACCESSTOKEN refers to an APIKEY with enough permissions at system/admin level.

Song can use whether Client credentials or an ApiKey.
However, as for now the only practical way is to use ApiKey assigned to an admin user. 

The reason is because we cannot assign policy/permissions on a Keycloak Client directly as these policy/permissions are usually assigned to user groups. Still unsure if this a limitation of Keycloak or just a wrong configuration on our Keycloak realm.

In the contrary, EGO policies can be assigned to Users or Applications.

<details>

>   <summary><b>Click here for a detailed breakdown</b></summary>

Reviewer: In the following section we are inconsistent about our sentence formatting. Bullet points should still have full sentences that end in periods. Many of these end without the period.

| Profile       | Description                                                                                                                                                                                                 |
|---------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `prod`        | Optimized for production use with minimal initialization and direct database connections.                                                                                                                                                   |
| `secure`      | Focuses on security with OAuth2 JWTs for API protection. This profile calls for a public key location for JWT verification and an introspection URI for authenticating clients.                                                                   |

> | `kafka`       | Targets Kafka integration, specifying Kafka bootstrap servers and a default topic for message exchange. Does not include specific configurations for other services or security settings.                               |

Reviewer: Enables Kafka integration.

Without this profile we don't send notifications.

Author: :D less stress

#### Flyway Variables

- `SPRING_FLYWAY_ENABLED` enables the initialization of the Song database with a Flyway database migration, setting up the necessary tables for API interactions. This migration utilizes SQL scripts located within Song and [found here](https://github.com/overture-stack/SONG/tree/develop/song-server/src/main/resources/db/migration). Without this setting, the database would remain uninitialized, leading to generic SQL errors (SQL Error: 0) with a SQLState of 42P01, corresponding to an undefined_table

#### Song Variables

> - `ID_USELOCAL` This mode indicates that Song will handle ID management internally, storing identifiers within its own system. Song can also be configured to use external ID management, for more information see our documentation for [ID management in Song](https://www.overture.bio/documentation/song/installation/configuration/id/)

Reviewer: 👍 Great note.

This is the only bullet point in this details section that starts with a property name then a capitalized sentence, all other bullets use the property name in the sentence. You may want to edit to keep the formatting consistent.

- `AUTH_SERVER_SCOPE_STUDY_PREFIX` the prefix added to the scope claim in the token. Scopes define the level of access granted to the token holder. In this case, it indicates a specific type of access related to studies

- `AUTH_SERVER_SCOPE_STUDY_SUFFIX` the suffix added to the scope claim in the token, further defining the level of access. Here, it specifies write access to study-related resources

- `AUTH_SERVER_SCOPE_SYSTEM` is the scope for system-level permissions, indicating write access to system resources managed by the application

- `SPRING_SECURITY_OAUTH2_RESOURCESERVER_JWT_JWK_SET_URI` is the URL where the JSON Web Key Set (JWS) for the JWT tokens is located. This key set is used by the application to validate the signature of the JWT tokens issued by Keycloak

- `AUTH_SERVER_INTROSPECTIONURI`: the URL used by the application to check the validity of a token against the Keycloak server. Introspection allows the application to verify that a token has not been revoked or expired

#### PostgreSQL connection details

> - `SPRING_DATASOURCE_URL`, `SPRING_DATASOURCE_USERNAME`, `SPRING_DATASOURCE_PASSWORD` are the connection details for the PostgreSQL database. The value for the `SPRING_DATASOURCE_URL` needs to be appended with  `?stringtype=unspecified` (Song as it is coded requires string type to be unspecified to interact with JSONb columns)

Reviewer: Glad to see a comment about the stringtype BS.

#### Kafka Variables

> - `SPRING_KAFKA_BOOTSTRAP-SERVERS` and `SPRING_KAFKA_TEMPLATE_DEFAULT-TOPIC` specifies the bootstrap servers and default topics for message publishing

Reviewer: This is super generic and I wish I could give more useful text but I would need to investigate our kafka setup more to understand the values here.

The Kafka Default Topic is relevant, as this is the topic in Kafka that Song will send messages to. Maestro must read from the same topic.

Author: Agreed Kafka information as a whole will need improvement, need to spent some time learning a little more and poking dev ops for information

#### Custom Swagger URL

- `SWAGGER_ALTERNATEURL` specifies an custome URL for accessing the Swagger UI (`/swagger-ui`)

<br></br>

---
</details>

<br></br>

2. **Run Song:** Use the docker run command with your `.env.song` file:

```bash
docker run -d \
  --name song \
  --platform linux/amd64 \
  -p 8080:8080 \
  --env-file .env.song \
  ghcr.io/overture-stack/song-server:438c2c42

> ```

Reviewer 2: we should change tag now to `latest`

AUTH_SERVER_URL=http://keycloak:8080/realms/myrealm/apikey/check_api_key/      
AUTH_JWT_PUBLICKEYURL=http://keycloak:8080/oauth/token/public_key
SPRING_SECURITY_OAUTH2_RESOURCESERVER_JWT_JWK_SET_URI=http://keycloak:8080/realms/myrealm/protocol/openid-connect/certs
```

<br></br>
<details>
  <summary><b>Click here for a detailed breakdown</b></summary>
<br></br>

#### Spring Run Profiles

- **Spring Run Profiles** activates specific profiles for the application with defined configurations. Profiles and their specified enviorment variables are defined in the [Score server application.yml](https://github.com/overture-stack/score/blob/develop/score-server/src/main/resources/application.yml). The profiles used here are summarized below

| Profile       | Description                                                                                                                                                                                                 |
|---------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `collaboratory` | Configures the service for use with the Cancer Collaboratory, including S3 endpoint, metadata server URL, and legacy mode settings.                                                            |

Reviewer: This one pains me, as Cancer Collaboratory no longer exists but this profile is MANDATORY for an S3 Score. Can we instead describe this as the profile to indicate to Score to use an S3 backend? It is aliased to `collaboratory` for legacy reasons, we will update this in the near future.

If you do want to name drop Cancer Collaboratory you should provide some context and a link ( https://cancercollaboratory.org/ ) which helpfully states  that it is now closed.

Author: Done!

| `prod`        | Optimizes the service for production use, enabling S3 security features and specifying the metadata server URL.                                                                                           |
| `secure`      | Implements OAuth authentication, specifying the authentication server URL, token name, client ID, client secret, and scopes for download and upload operations.                                  |
| `JWT`      |  The JWT (JSON Web Token) profile is used to configure the authentication method based on JWT. This profile includes settings to obtain the public key for token validation from an OAuth server.                             |

#### Song & Score Variables

>   - `SERVER_PORT` and `SERVER_SSL_ENABLED` specifies the port for the Score service (`8087`) and disables SSL (`false`), indicating HTTP communication.

Reviewer: You may want a note here that SSL disabling is for development setup only, to avoid needing to configure SSL Certificates for HTTPS.

---
title: Deployment Guide
---

# Data Management and Storage Setup

> Song, Score and their dependent services (Postgres, Minio, and Kafka) will be set up next. These services will manage, track and store all our data on the backend. 

Reviewer 3: On the same "consistent spelling check" spirit as my previous comment: backend vs back-end vs back end...

```suggestion
Song, Score and the services they depend on (Postgres, Minio, and Kafka) will be set up next. These Overture services will manage, track and store all our data on the backend. 
```

> Song, Score and the services they depend on (Postgres, Minio, and Kafka) will be set up next. These Overture services will manage, track and store all our data on the backend. 

Reviewer 3: ```suggestion
Song, Score and the services they depend on (Postgres, MinIO, and Kafka) will be set up next. These Overture services will manage, track and store all our data on the backend. 
```

## Setting up Object Storage

> The file transfer service Score is compatible with any S3 storage provider; for simplicity, we will use the open-source object Storage provider Minio for this setup. 

Reviewer 3: ```suggestion
The file transfer service Score is compatible with any S3 storage provider; for simplicity, we will use the open-source object Storage provider MinIO for this setup. 
```

> 1. **Run Minio:** Use the following command to pull and run the Minio docker container

Reviewer 3: ```suggestion
1. **Run MinIO:** Use the following command to pull and run the MinIO docker container
```

```bash
docker run -d --name minio \
  -p 9000:9000 \
  -e MINIO_ACCESS_KEY=admin \
  -e MINIO_SECRET_KEY=admin123 \
  -v $./persistentStorage/data-minio:/data \
  minio/minio:RELEASE.2018-05-11T00-29-24Z \
  server /data
```

> **Create buckets in Minio:** Run the following command to create two buckets in the running Minio server using the Minio CLI - this will create an **object** bucket to store uploaded files and a **state** bucket for metadata files managed by Score.

Reviewer 3: ```suggestion
**Create buckets in MinIO:** Run the following command to create two buckets in the running MinIO server using the MinIO CLI - this will create an **object** bucket to store uploaded files and a **state** bucket for metadata files managed by Score.
```

```bash
docker run --name minio-client \
  --entrypoint /bin/sh \
  minio/mc -c \
  '/usr/bin/mc alias set myminio http://host.docker.internal:9000 admin admin123 && \
  /usr/bin/mc mb myminio/state && \
  /usr/bin/mc mb myminio/object && \
  /usr/bin/mc mb myminio/object/data && \
  echo "" > /tmp/heliograph && \
  /usr/bin/mc put /tmp/heliograph myminio/object/data/heliograph && \
  rm /tmp/heliograph && \
  exit 0;'
```

You should now be able to access the minio console from the browser at `localhost:9000`

<details>

  <summary><b>For more details, click here</b></summary>

<br></br>

### Minio Image

> - The `-v $./persistentStorage/data-minio:/data` configures Minio to store data in our local filesystem instead of in the docker container. Files you upload to Minio will be stored at the path `./persistentStorage/data-minio`.

Reviewer 3: ```suggestion
- The `-v $./persistentStorage/data-minio:/data` configures MinIO to store data in our local file system instead of in the docker container. Files you upload to MinIO will be stored at the path `./persistentStorage/data-minio`.
```

### Minio Client Image

> - **Alias:** `alias set myminio http://host.docker.internal:9000 admin admin123` creates an `alias` for the Minio server, with an `admin` user with a the password `admin123`.

Reviewer 3: ```suggestion
- **Alias:** `alias set myminio http://host.docker.internal:9000 admin admin123` creates an `alias` for the MinIO server, with an `admin` user with a the password `admin123`.
```

---

</details>

<br></br>

## Setting up the Song Database

1. **Run PostgreSQL:** Use the following command to pull and run the PostgreSQL docker container

```bash
docker run --name song-db \
 -e POSTGRES_USER=admin \
 -e POSTGRES_PASSWORD=admin123 \
 -e POSTGRES_DB=songDb 
 -v ./persistentStorage/data-song-db:/var/lib/postgresql/data \
-d postgres:11.1 
```
<details>

  <summary><b>For more details, click here</b></summary>

<br></br>

- This command runs a postgres image named `song-db` with the username `admin`, password `admin123` and a database within it called `songDb`.

- We have included a persistent volume `-v ./persistentStorage/song-db-data:/var/lib/postgresql/data`/. This volume stores Songs Postgres data in our local filesystem instead of the docker container; in other words, the data contained in the Song database will be stored at the path `./persistentStorage/song-db-data:/var/lib/postgresql/data`.

---
</details>

<br></br>

> <Note title="Bringing it together">By setting up Minio alongside PostgreSQL, we have created an environment capable of handling both relational and object data storage, simulating a backend infrastructure similar to what you'd find in a cloud-based application.</Note>

Reviewer 3: ```suggestion
<Note title="Bringing it together">By setting up MinIO alongside PostgreSQL, we have created an environment capable of handling both relational and object data storage, simulating a backend infrastructure similar to what you'd find in a cloud-based application.</Note>
```
